### ISSUE TITLE
Python API pending features

### ISSUE THREAD

**Issue by ZJaume** [ZJaume/heliport#6]
- [x] Expose identifier parameters as python arguments.
- ~~[ ] Method to download models programatically~~
- [x] Batched multi-threaded prediction.
- [x] Internal static shared Model to allow multiple Python Identifier objects sharing the same data and not allocating a whole new one each time. Should allow multiple Python Identifiers in separated threads work fast.
- [x] Return confidence scores when requested.
- [x] Collapse langs in topk.

**Comment by svirpioj** [comment]
Hi, I'm planning to include heliport as one alternative language identifier to [OpusFilter](https://github.com/Helsinki-NLP/OpusFilter). Apart from the identifier parameters, it would be great to have the confidence values in the output. (`Identifier` seems to already have undocumented `ignore_confidence` argument, but it has no observable effect.)

**Comment by ZJaume** [comment]
The `ignore_confidence` will do the same as it is documented in the CLI, ignoring confidence threshold when determining the language, so texts under the threshold won't get `und` as label.

Implementing all these options are in my plans, but haven't had time to do it yet.

**Comment by ZJaume** [comment]
@svirpioj a new version has been published including more python functions including the feature you needed.

**Comment by svirpioj** [comment]
> [@svirpioj](https://github.com/svirpioj) a new version has been published including more python functions including the feature you needed.

Thanks! The `ignore_confidence` setting does not seems to affect `identify_topk_with_score` output, though:
```python
import heliport
i_f = heliport.Identifier(ignore_confidence=False)
i_t = heliport.Identifier(ignore_confidence=True)

i_t.identify_with_score("he")
# ('cho', 1.6336495876312256)
i_f.identify_with_score("he")
# ('und', 0.1509031057357788)

i_t.identify_topk_with_score("he", k=3)
# [('cho', 1.6336495876312256), ('mri', 1.7845526933670044), ('gom', 2.5137245655059814)]
i_f.identify_topk_with_score("he", k=3)
# [('cho', 1.6336495876312256), ('mri', 1.7845526933670044), ('gom', 2.5137245655059814)]
```

Also an option for the relevant languages would be nice to have.

**Comment by ZJaume** [comment]
> Edit: Well, probably that's a feature and not a bug (difference with the 2nd best doesn't make that much sense in the top k list), but in that case the docstring of the method is a bit confusing.

You are right, the topk scoring does not change because confidence is more or less meaningless. Maybe the whole `*_with_score` stuff needs another rework. I don't have a clear picture of which would be the best API. Like in the CLI, I was trying to be consistent with the original implementation: using an option to show scores and switch between confidence and raw score with another option.

> Edit 2: And for the raw scores, lower is better, right? The README and command line help tell otherwise.

I'm fixing it.

> Edit 3: For the output of identify_topk_with_score methods, I think the collapse function for the languages would also be useful.

Will add this as a TODO

> Also an option for the relevant languages would be nice to have.
Unfortunately this is rather difficult to have, as it needs the language model text format which is not being included in the Python binary package. I will need to take a look on how to implement it and may take time.

**Comment by ZJaume** [comment]
Ok, so here is a sketch of a possible rework of the Python API. Disabling confidence via parameter on each call, I think it will be safer for client code.
```python
# Identify the language of a string.
#
# If confidence threshold is enabled (default), all predictions below
# the threshold will be labeled as 'und'.
def identify(text: str, conf_threshold: bool) -> str

# Identify the language of a string and return the prediction score.
#
# When confidence threshold is enabled (default), all predictions below
# the threshold will be labeled as 'und'. The returned score is the confidence
# value (higher is better) if the threshold is enabled, and the
# raw score if the threshold is disabled.
def identify_score(text: str, conf_threshold: bool) -> (str, float)

# Identify the top-k languages of a string.
def identify_topk(text: str, k: int) -> list[str]

# Identify the top-k languages of a string and return each language raw score.
def identify_topk_score(text: str, k: int) -> list[(str, float)]

# Parallelized version of `identify`, which takes a batch of strings
# and runs the identification in parallel.
def par_identify(texts: list[str], conf_threshold: bool) -> list[str]

# Parallelized version of `identify_score`, which takes a batch of strings
# and runs the identification in parallel.
def par_identify_score(texts: list[str], conf_threshold: bool) -> list[(str, float)]

# Get the confidence threshold for a language
def confidence_threshold(lang: str) -> float
```

**Comment by svirpioj** [comment]
Looks good to me.
